clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java

```diff
@@ -207,8 +207,7 @@ private void reapExpiredApplicationEvents(long currentTimeMs) {
      */
     // Visible for testing
     static void runAtClose(final Collection<Optional<? extends RequestManager>> requestManagers,
-                           final NetworkClientDelegate networkClientDelegate,
-                           final Timer timer) {
```

Comment:
I guess this timer was expected to be used by `pollOnClose`. It seems to me `pollOnClose` should take a `timeout`, so the request managers has the chance to return on time in the future.

For another, the `timer` should call `timer.update()` at the end. Otherwise, the later operations (`sendUnsentRequests`) will get incorrect remaining timeout.

Reply:
Agreed, that was probably the thinking behind the timer here, but it ended up unused and that actually seemed sensible to me here (because it's used somewhere else). This is how I see it: `pollOnClose` is only building requests without blocking on any activity that would need a time boundary (just like poll() doesn't take a timer to limit its execution). It's then the `sendUnsentRequests` the one that takes the requests created by pollOnClose and performs the operation that needs the timer (and it has it already). Taking the pollOnClose on the FetchRequestManager as example (actually the only manager where it's used), it only generates a fetch request to close the fetch sessions (no timer needed). Makes sense?

Reply:
thanks for sharing, and I agree that the usage of `pollOnClose` you described, so +1 to remove `timer` :)